modules/service/src/main/scala/lucuma/odb/service/GroupService.scala

```diff
@@ -62,6 +71,68 @@ object GroupService {
           r.traverse(createGroupImpl(_, input.SET, input.initialContents, system))
 
+      // Clone `oid` into `dest`, at the end.
+      private def cloneObservationInto(oid: Observation.Id, dest: Option[Group.Id])(using Transaction[F]): ResultT[F, Observation.Id] =
+        ResultT(
+          observationService.cloneObservation(
+            CloneObservationInput(
+              observationId  = Some(oid),
+              observationRef = None,
+              SET = Some(
+                ObservationPropertiesInput.Edit.Empty.copy(
+                  group = Nullable.orNull(dest)
+                )
+              )
+            )
+          )
+        ).map(_.cloneId)
+
+      // Clone `gid` into `dest`, at the end, as an empty group.
+      private def cloneAsEmptyGroupInto(gid: Group.Id, dest: Option[Group.Id])(using Transaction[F]): ResultT[F, Group.Id] =
+        selectGroupAsInput(gid).flatMap: input => 
+          ResultT(createGroup(input.copy(SET = input.SET.copy(parentGroupId = dest, parentGroupIndex = None))))
+        
+      // Clone `gid` into `dest`, at the end, and clone its contents too.
+      private def cloneGroupInto(gid: Group.Id, dest: Option[Group.Id])(using Transaction[F]): ResultT[F, Group.Id] =
+        for
+          gid0 <- cloneAsEmptyGroupInto(gid, dest)
+          es   <- selectGroupElements(gid) // find the children of the source group
+          _    <- es.traverse(cloneGroupElementInto(_, Some(gid0))) // clone the children, recursively, into the new group
+        yield gid
+
+      // Clone the `elem` into `dest`, at the end.
+      private def cloneGroupElementInto(elem: GroupElement.Id, dest: Option[Group.Id])(using Transaction[F]): ResultT[F, GroupElement.Id] =
+        elem match
+          case Left(gid)  => cloneGroupInto(gid, dest).map(_.asLeft)
+          case Right(oid) => cloneObservationInto(oid, dest).map(_.asRight)
+
+      /** Construct a CreateGroupInput that would clone `gid`. */
+      private def selectGroupAsInput(gid: Group.Id): ResultT[F, CreateGroupInput] =
+        ResultT:
+          session.prepareR(Statements.SelectGroupAsInput).use: pq =>
+            pq.unique(gid).map:
+              case (cgi, false) => Result.success(cgi)
+              case (_, true)    => OdbError.UpdateFailed(Some("System groups cannot be cloned.")).asFailure
+
+      // Select the elements of `gid`, in order.
+      private def selectGroupElements(gid: Group.Id): ResultT[F, List[GroupElement.Id]] =
+        ResultT:
+          session.prepareR(Statements.SelectGroupElements).use: pq =>
+            pq.stream(gid, 1024).compile.toList.map(Result.success)
+
+      // Clone `gid` as a sibling.
+      private def cloneGroupImpl(input: CloneGroupInput)(using Transaction[F]): ResultT[F, Group.Id] =
+        for
+          cgi   <- selectGroupAsInput(input.groupId)
+          cgiʹ   = cgi.copy(SET = input.SET.foldLeft(cgi.SET)(_.withEdit(_)))
+          clone <- ResultT(createGroup(cgiʹ, false))
```

Comment:
Is the sibling added at the end or next door?

Reply:
The clone is inserted at the specified index, or at the source index if left unspecified (in which case the source will be pushed down to make room).